### ISSUE TITLE
PDF & DOCX Export

### ISSUE THREAD

**Issue by mlava** [roamhacker/SmartBlocks#209]
## ✂️ Copy of your #42SmartBlock from Roam

[PDF DOCX Export.zip](https://github.com/roamhacker/SmartBlocks/files/5987334/PDF.DOCX.Export.zip)

## 📋 Describe the SmartBlock

This SmartBlock exports the current page as either PDF or DOCX format.

You can configure whether you want to include the linked references section from the bottom of the page.

It relies on @roamhacker's [Roam42](https://roamresearch.com/#/app/roamhacker/page/jI-X_cwaf) conversion engine, using the Markdown flattening and Markdown to HTML functions.

Note: all the available scripts to convert html to pdf do so by creating an image and then printing that to pdf. This will mean your pdf is _not_ searchable. I will be happy to update if/when someone smarter than me figures out how to export html to pdf without doing so via image.

Created to fulfil a bounty posted by Justin Shaffner and the linked references section by Owen Cyrulnik.

## ✅ Describe any prerequisites or dependencies that are required for this SmartBlock

Roam 42 and roam/js.

## 💡 Additional Info

Loom video: https://www.loom.com/share/23a3e2f913cc4cb69aff548fe10a56a6

**Comment by stevenhill1** [comment]
This is fantastically useful - thank you. I don't have the expertise, but would it be feasible to extend this using https://docverter.com/api/

**Comment by migdur** [comment]
Hi, sorry, pretty new to smartblocks here. Am I supposed to copy /paste the contents of PDF DOCX Export.zip into a block in a Smartblock workflow? This does not seem to work for me. Thanks for helping!

**Comment by stevenhill1** [comment]
> Hi, sorry, pretty new to smartblocks here. Am I supposed to copy /paste the contents of PDF DOCX Export.zip into a block in a Smartblock workflow? This does not seem to work for me. Thanks for helping!

You need to extract the json file from the zip archive and then import the file (PDF & DOCX Export.json) into your Roam database (choose 'import files' from the ... menu). This will add a new page to you Roam graph which has full instructions for set up.

**Comment by mlava** [comment]
@stevenhill1 Thanks, I am glad it's helpful. Will take a look at your link. And thanks for helping @migdur.

@migdur - did you get it working?

**Comment by mlava** [comment]
New version in first post.

Fixed: removal of trigger phrase before rendering of pdf/docx file.

**Comment by migdur** [comment]
> @stevenhill1 Thanks, I am glad it's helpful. Will take a look at your link. And thanks for helping @migdur.
> 
> @migdur - did you get it working?

Yes I did - thanks!

**Comment by lkonstan** [comment]
This is great; would love it if it were somehow able to parse and output footnotes.

**Comment by mottonew** [comment]
This looks very useful. I've installed, fire the SB, but it just disappears without exporting anything. That goes for both the PDF and the DOCX formats. It looks like the Loom is on a PC and I'm on a Mac - would that affect the export? I'm using Roam on Chrome, which is set to export to /downloads after asking. I've tried turning off other jss in Roam but that doesn't make any difference. My other SBs are working fine.
Thanks.

**Comment by mlava** [comment]
Did you uncomment the correct javascript imports in the roam/js block?

**Comment by ItsJamesWeiss** [comment]
Hey, I'm having trouble with the Docx exporting—they're blank when I try to open them. I'm on Mac btw.

**Comment by mottonew** [comment]
Yes, I uncommented all of them, removing the \\

Where would the exported files go - to where the setting on Chrome tells them?

Thanks for getting to me so quickly

**Comment by mlava** [comment]
Yes. You should get a download appear at the bottom of the window. Other
thing to do would be check downloads using ctrl-j.

**Comment by mlava** [comment]
@mottonew 
Could you send me a screenshot of your roam/js block please?

**Comment by cobblepot1** [comment]
This is great! My docx export did not preserve indentation - everything was at root level. Is this intended?

**Comment by mlava** [comment]
Hi. Yes, the original request from the person who posted the bounty was to
simplify (flatten) the markdown language for readability.
It can be modified, but we did find it was hard to follow in docx without
flattening.
Mark

**Comment by RoamStack** [comment]
Part of the SmartBlock trigger (jj) still shows up when I export to PDF (see first line in the attached example). When I export to .docx, the file is empty.

When installing the script, I didn't have to uncomment any line as they weren't commented:
![image](https://user-images.githubusercontent.com/77009918/109800536-94f65700-7c1d-11eb-814a-d7ee8a0ae7b4.png)

PDF (see jj in the first line):
[Export pages using Roam42.pdf](https://github.com/roamhacker/SmartBlocks/files/6075662/Export.pages.using.Roam42.pdf)

DOCX (completely empty):
[Export pages using Roam42.docx](https://github.com/roamhacker/SmartBlocks/files/6075663/Export.pages.using.Roam42.docx)

**Comment by mlava** [comment]
Can you check which version of the script you're using?
Current is 0.2

**Comment by orkhan10** [comment]
For this page, which is rather short, Docx is empty, pdf is not.
[Roman Empire (1).docx](https://github.com/roamhacker/SmartBlocks/files/6095267/Roman.Empire.1.docx)
[Roman Empire (1).pdf](https://github.com/roamhacker/SmartBlocks/files/6095268/Roman.Empire.1.pdf)


I have a longer page. PDF is empty but 50 pages long. Docx is also empty but only 1 page long.

[Course_Stock 1_ Foundation (1).docx](https://github.com/roamhacker/SmartBlocks/files/6095274/Course_Stock.1_.Foundation.1.docx)
[Course_Stock 1_ Foundation (4).pdf](https://github.com/roamhacker/SmartBlocks/files/6095275/Course_Stock.1_.Foundation.4.pdf)

And also is it possible to improve the image quality of pdf? It is actually a bit interesting. Blurry on 100% zoom but better when zoomed in.

**Comment by mlava** [comment]
@RoamStack 
Are you on Mac by any chance?

@orkhan10 appears to be on Mac and is having issues.

I wonder if it is an OS issue as I am using Windows and both pdf and docx are working perfectly, without the trigger phrase and no empty files.

**Comment by stevenhill1** [comment]
> @RoamStack
> Are you on Mac by any chance?
> 
> @orkhan10 appears to be on Mac and is having issues.
> 
> I wonder if it is an OS issue as I am using Windows and both pdf and docx are working perfectly, without the trigger phrase and no empty files.

Also works fine for me on Windows, but getting these problems on Safari on iOS

**Comment by mlava** [comment]
Ok, well that's interesting.
I don't have a Mac or iphone so can't troubleshoot.
Mark

**Comment by diegouriarte** [comment]
Hi, in mi case it leaves jjdoc in exported document, and it doesn't recognizes Spanish characters 
![image](https://user-images.githubusercontent.com/678268/112547088-481f1f80-8d88-11eb-8640-18cae32cbb2d.png)

**Comment by edumanos** [comment]
> Hi, in mi case it leaves jjdoc in exported document, and it doesn't recognizes Spanish characters
> ![image](https://user-images.githubusercontent.com/678268/112547088-481f1f80-8d88-11eb-8640-18cae32cbb2d.png)

Turkish characters either :(

**Comment by sandeepkelvadi** [comment]
![image](https://user-images.githubusercontent.com/85609774/121313891-75f61800-c924-11eb-9cd1-f51c55843d04.png)
The trigger doesn't seem to do anything for me. I'm on Windows and using the Brave browser. Also tried on Chrome, but no luck. Is there something I'm missing. I followed the instructions in the configuration section.

**Comment by zzzgithubzzz** [comment]
Hi, 
This is a great utility. Is there a way to export the page along with all the images embedded within a page while exporting as pdf with this smartblock trigger?
Vinod

**Comment by jbnv** [comment]
I have custom CSS in my graph for both light and dark mode. I usually work in dark mode, but I definitely don't want the dark mode exported to PDF. Is there a way around this?

**Comment by Elwood4415** [comment]
This is a tremendous resource, but from time-to-time it quits working. Usually, restarting Roam gets it working again and if that doesn't then rebooting the computer does. But now it is not working at all. As far as I can tell, everything is still present. Any tips would be appreciated.

**Comment by mlava** [comment]
The DOCX version of this script is now available in the SmartBlocks store within Roam Research.

The required roam/js to allow it to work is below:

`- {{roam/js}}
    - ```javascript
function addScripts(url) {
   var script = document.createElement('script');
   script.type = 'application/javascript';
   script.src = url;
   document.head.appendChild(script);
}

addScripts('https://evidenceprime.github.io/html-docx-js/build/html-docx.js');
addScripts('https://cdnjs.cloudflare.com/ajax/libs/FileSaver.js/1.3.8/FileSaver.js');
addScripts('https://tinymce.cachefly.net/4.1/tinymce.min.js');```
`

**Comment by mlava** [comment]
The PDF version of this script is now available in the SmartBlocks store within Roam Research.

The required roam/js to allow it to work is below:

- {{roam/js}}
    - ```javascript    function addScripts(url) {
       var script = document.createElement('script');
       script.type = 'application/javascript';
       script.src = url;
       document.head.appendChild(script); }     addScripts('https://raw.githack.com/eKoopmans/html2pdf/master/dist/html2pdf.bundle.js');```
